File: kuber-client-java/examples/IbmMqRequestResponseTest.java
/*
 * Copyright © 2025-2030, All Rights Reserved
 * Ashutosh Sinha | Email: ajsinha@gmail.com
 *
 * Kuber IBM MQ Request/Response Test Client (v1.7.5)
 * 
 * Tests the Kuber Request/Response messaging feature via IBM MQ.
 *
 * Prerequisites:
 * - IBM MQ running with queue manager accessible
 * - Kuber server running with messaging enabled
 * - IBM MQ JMS client libraries:
 *     <dependency>
 *         <groupId>com.ibm.mq</groupId>
 *         <artifactId>com.ibm.mq.allclient</artifactId>
 *         <version>9.3.4.0</version>
 *     </dependency>
 */
package com.kuber.client.examples;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.ibm.msg.client.jms.JmsConnectionFactory;
import com.ibm.msg.client.jms.JmsFactoryFactory;
import com.ibm.msg.client.wmq.WMQConstants;

import javax.jms.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * Test client for Kuber's IBM MQ-based Request/Response messaging.
 * 
 * @version 1.7.5
 */
public class IbmMqRequestResponseTest {
    
    // Configuration
    private static final String QUEUE_MANAGER = "QM1";
    private static final String CHANNEL = "DEV.APP.SVRCONN";
    private static final String HOST = "localhost";
    private static final int PORT = 1414;
    
    private static final String REQUEST_QUEUE = "ccs_cache_request";
    private static final String RESPONSE_QUEUE = "ccs_cache_response";
    
    // Your Kuber API key
    private static final String API_KEY = "kub_your_api_key_here";
    private static final String TEST_REGION = "default";
    
    // JSON mapper
    private final ObjectMapper objectMapper = new ObjectMapper();
    
    // JMS objects
    private Connection connection;
    private Session session;
    private MessageProducer producer;
    private MessageConsumer consumer;
    
    // State
    private final AtomicBoolean running = new AtomicBoolean(true);
    private final Map<String, CompletableFuture<JsonNode>> pendingRequests = new ConcurrentHashMap<>();
    private final List<JsonNode> receivedResponses = new CopyOnWriteArrayList<>();
    
    public static void main(String[] args) {
        System.out.println("╔════════════════════════════════════════════════════════════════════╗");
        System.out.println("║  Kuber IBM MQ Request/Response Test Client (Java)                  ║");
        System.out.println("║  Tests cache operations via IBM MQ messaging                       ║");
        System.out.println("╚════════════════════════════════════════════════════════════════════╝");
        System.out.println();
        
        IbmMqRequestResponseTest client = new IbmMqRequestResponseTest();
        try {
            client.run();
        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
        }
    }
    
    public void run() throws Exception {
        System.out.println("Queue Manager: " + QUEUE_MANAGER);
        System.out.println("Channel: " + CHANNEL);
        System.out.println("Host: " + HOST + ":" + PORT);
        System.out.println("Request Queue: " + REQUEST_QUEUE);
        System.out.println("Response Queue: " + RESPONSE_QUEUE);
        System.out.println();
        
        try {
            connect();
            startResponseListener();
            
            Thread.sleep(1000);
            
            runTestSuite();
            
            System.out.println("\nWaiting for responses (15 seconds)...");
            Thread.sleep(15000);
            
        } finally {
            shutdown();
            printSummary();
        }
        
        System.out.println("\nTest complete.");
    }
    
    private void connect() throws Exception {
        System.out.println("Connecting to IBM MQ...");
        
        JmsFactoryFactory ff = JmsFactoryFactory.getInstance(WMQConstants.WMQ_PROVIDER);
        JmsConnectionFactory cf = ff.createConnectionFactory();
        
        cf.setStringProperty(WMQConstants.WMQ_HOST_NAME, HOST);
        cf.setIntProperty(WMQConstants.WMQ_PORT, PORT);
        cf.setStringProperty(WMQConstants.WMQ_CHANNEL, CHANNEL);
        cf.setIntProperty(WMQConstants.WMQ_CONNECTION_MODE, WMQConstants.WMQ_CM_CLIENT);
        cf.setStringProperty(WMQConstants.WMQ_QUEUE_MANAGER, QUEUE_MANAGER);
        
        connection = cf.createConnection();
        session = connection.createSession(false, Session.AUTO_ACKNOWLEDGE);
        
        Queue requestQueue = session.createQueue("queue:///" + REQUEST_QUEUE);
        Queue responseQueue = session.createQueue("queue:///" + RESPONSE_QUEUE);
        
        producer = session.createProducer(requestQueue);
        producer.setDeliveryMode(DeliveryMode.PERSISTENT);
        
        consumer = session.createConsumer(responseQueue);
        
        connection.start();
        System.out.println("✓ Connected to IBM MQ");
    }
    
    private void startResponseListener() throws JMSException {
        consumer.setMessageListener(message -> {
            try {
                if (message instanceof TextMessage) {
                    String responseText = ((TextMessage) message).getText();
                    JsonNode response = objectMapper.readTree(responseText);
                    
                    // Parse nested response
                    JsonNode requestNode = response.get("request");
                    JsonNode responseNode = response.get("response");
                    
                    String messageId = requestNode != null && requestNode.has("message_id") ?
                        requestNode.get("message_id").asText() : "unknown";
                    boolean success = responseNode != null && responseNode.has("success") &&
                        responseNode.get("success").asBoolean();
                    String operation = requestNode != null && requestNode.has("operation") ?
                        requestNode.get("operation").asText() : "unknown";
                    long processingTime = response.has("processing_time_ms") ?
                        response.get("processing_time_ms").asLong() : 0;
                    
                    System.out.println("\n════════════════════════════════════════════════════════════");
                    System.out.println("📥 RESPONSE RECEIVED");
                    System.out.println("   Message ID: " + messageId);
                    System.out.println("   Success: " + success);
                    System.out.println("   Operation: " + operation);
                    System.out.println("   Processing Time: " + processingTime + "ms");
                    
                    if (success) {
                        if (responseNode != null && responseNode.has("result")) {
                            String result = responseNode.get("result").toString();
                            if (result.length() > 100) {
                                System.out.println("   Result: " + result.substring(0, 100) + "...");
                            } else {
                                System.out.println("   Result: " + result);
                            }
                        }
                    } else {
                        if (responseNode != null) {
                            String errorCode = responseNode.has("error_code") ?
                                responseNode.get("error_code").asText() : "UNKNOWN";
                            String error = responseNode.has("error") ?
                                responseNode.get("error").asText() : "Unknown error";
                            System.out.println("   Error Code: " + errorCode);
                            System.out.println("   Error: " + error);
                        }
                    }
                    System.out.println("════════════════════════════════════════════════════════════");
                    
                    receivedResponses.add(response);
                    
                    CompletableFuture<JsonNode> future = pendingRequests.remove(messageId);
                    if (future != null) {
                        future.complete(response);
                    }
                }
            } catch (Exception e) {
                System.err.println("Error processing response: " + e.getMessage());
            }
        });
        
        System.out.println("✓ Response listener started");
    }
    
    private void runTestSuite() throws Exception {
        System.out.println("\n════════════════════════════════════════════════════════════");
        System.out.println("STARTING TEST SUITE");
        System.out.println("════════════════════════════════════════════════════════════");
        
        // Test 1: SET
        System.out.println("\n--- Test 1: SET operation ---");
        ObjectNode setRequest = createRequest("SET", TEST_REGION, "ibmmq-test:user:1001");
        ObjectNode userData = objectMapper.createObjectNode();
        userData.put("name", "John Doe");
        userData.put("email", "john@example.com");
        userData.put("broker", "ibmmq");
        setRequest.set("value", userData);
        setRequest.put("ttl", 300);
        publishRequest(setRequest);
        Thread.sleep(1000);
        
        // Test 2: GET
        System.out.println("\n--- Test 2: GET operation ---");
        publishRequest(createRequest("GET", TEST_REGION, "ibmmq-test:user:1001"));
        Thread.sleep(1000);
        
        // Test 3: EXISTS
        System.out.println("\n--- Test 3: EXISTS operation ---");
        publishRequest(createRequest("EXISTS", TEST_REGION, "ibmmq-test:user:1001"));
        Thread.sleep(1000);
        
        // Test 4: SET another
        System.out.println("\n--- Test 4: SET another key ---");
        ObjectNode setRequest2 = createRequest("SET", TEST_REGION, "ibmmq-test:user:1002");
        ObjectNode userData2 = objectMapper.createObjectNode();
        userData2.put("name", "Jane Smith");
        userData2.put("role", "admin");
        setRequest2.set("value", userData2);
        publishRequest(setRequest2);
        Thread.sleep(1000);
        
        // Test 5: KEYS
        System.out.println("\n--- Test 5: KEYS operation ---");
        publishRequest(createRequest("KEYS", TEST_REGION, "ibmmq-test:*"));
        Thread.sleep(1000);
        
        // Test 6: DELETE
        System.out.println("\n--- Test 6: DELETE operation ---");
        publishRequest(createRequest("DELETE", TEST_REGION, "ibmmq-test:user:1002"));
        Thread.sleep(1000);
        
        // Test 7: GET non-existent
        System.out.println("\n--- Test 7: GET non-existent key ---");
        publishRequest(createRequest("GET", TEST_REGION, "ibmmq-test:nonexistent:key"));
        Thread.sleep(1000);
        
        System.out.println("\n✓ All requests sent");
    }
    
    private ObjectNode createRequest(String operation, String region, String key) {
        String messageId = "req-" + UUID.randomUUID().toString().substring(0, 12);
        
        ObjectNode request = objectMapper.createObjectNode();
        request.put("operation", operation);
        request.put("region", region);
        request.put("key", key);
        request.put("api_key", API_KEY);
        request.put("message_id", messageId);
        
        pendingRequests.put(messageId, new CompletableFuture<>());
        return request;
    }
    
    private void publishRequest(ObjectNode request) throws Exception {
        String messageId = request.get("message_id").asText();
        String requestJson = objectMapper.writeValueAsString(request);
        
        System.out.println("📤 SENDING REQUEST");
        System.out.println("   Message ID: " + messageId);
        System.out.println("   Operation: " + request.get("operation").asText());
        System.out.println("   Key: " + request.get("key").asText());
        
        TextMessage message = session.createTextMessage(requestJson);
        message.setJMSCorrelationID(messageId);
        producer.send(message);
        
        System.out.println("   ✓ Sent");
    }
    
    private void shutdown() {
        System.out.println("\nShutting down...");
        running.set(false);
        
        try {
            if (producer != null) producer.close();
            if (consumer != null) consumer.close();
            if (session != null) session.close();
            if (connection != null) connection.close();
        } catch (Exception e) {
            System.err.println("Error closing connections: " + e.getMessage());
        }
    }
    
    private void printSummary() {
        System.out.println("\n════════════════════════════════════════════════════════════");
        System.out.println("TEST SUMMARY");
        System.out.println("════════════════════════════════════════════════════════════");
        System.out.println("Total responses received: " + receivedResponses.size());
        System.out.println("Pending requests (no response): " + pendingRequests.size());
        
        if (!pendingRequests.isEmpty()) {
            System.out.println("\nRequests without responses:");
            for (String messageId : pendingRequests.keySet()) {
                System.out.println("  - " + messageId);
            }
        }
        
        long successCount = receivedResponses.stream()
            .filter(r -> {
                JsonNode responseNode = r.get("response");
                return responseNode != null && responseNode.has("success") && 
                    responseNode.get("success").asBoolean();
            })
            .count();
        long errorCount = receivedResponses.size() - successCount;
        
        System.out.println("\nSuccessful operations: " + successCount);
        System.out.println("Failed operations: " + errorCount);
        System.out.println("════════════════════════════════════════════════════════════");
    }
}

File: kuber-client-csharp/examples-standalone/IbmMqRequestResponseTest.cs
/*
 * Copyright © 2025-2030, All Rights Reserved
 * Ashutosh Sinha | Email: ajsinha@gmail.com
 *
 * Kuber IBM MQ Request/Response Test Client (v1.7.5)
 * 
 * Tests the Kuber Request/Response messaging feature via IBM MQ.
 * 
 * Prerequisites:
 * - IBM MQ running with queue manager accessible
 * - Kuber server running with messaging enabled
 * - IBMMQDotnetClient NuGet package:
 *     dotnet add package IBMMQDotnetClient
 */

using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Threading;
using System.Threading.Tasks;
using IBM.WMQ;

namespace Kuber.Client.Examples
{
    /// <summary>
    /// Test client for Kuber's IBM MQ-based Request/Response messaging.
    /// </summary>
    public class IbmMqRequestResponseTest
    {
        // Configuration
        private const string QueueManagerName = "QM1";
        private const string ChannelName = "DEV.APP.SVRCONN";
        private const string HostName = "localhost";
        private const int Port = 1414;
        
        private const string RequestQueueName = "ccs_cache_request";
        private const string ResponseQueueName = "ccs_cache_response";
        
        // Your Kuber API key
        private const string ApiKey = "kub_your_api_key_here";
        private const string TestRegion = "default";
        
        // MQ objects
        private MQQueueManager? _queueManager;
        private MQQueue? _requestQueue;
        private MQQueue? _responseQueue;
        
        // State
        private readonly CancellationTokenSource _cts = new();
        private readonly ConcurrentDictionary<string, TaskCompletionSource<JsonNode>> _pendingRequests = new();
        private readonly List<JsonNode> _receivedResponses = new();
        private readonly object _responseLock = new();
        
        public static async Task Main(string[] args)
        {
            Console.WriteLine("╔════════════════════════════════════════════════════════════════════╗");
            Console.WriteLine("║  Kuber IBM MQ Request/Response Test Client (C#)                    ║");
            Console.WriteLine("║  Tests cache operations via IBM MQ messaging                       ║");
            Console.WriteLine("╚════════════════════════════════════════════════════════════════════╝");
            Console.WriteLine();
            
            var client = new IbmMqRequestResponseTest();
            try
            {
                await client.RunAsync();
            }
            catch (Exception ex)
            {
                Console.Error.WriteLine($"Error: {ex.Message}");
                Console.Error.WriteLine(ex.StackTrace);
            }
        }
        
        public async Task RunAsync()
        {
            Console.WriteLine($"Queue Manager: {QueueManagerName}");
            Console.WriteLine($"Channel: {ChannelName}");
            Console.WriteLine($"Host: {HostName}:{Port}");
            Console.WriteLine($"Request Queue: {RequestQueueName}");
            Console.WriteLine($"Response Queue: {ResponseQueueName}");
            Console.WriteLine();
            
            try
            {
                Connect();
                var consumerTask = StartResponseConsumerAsync(_cts.Token);
                
                await Task.Delay(1000);
                
                await RunTestSuiteAsync();
                
                Console.WriteLine("\nWaiting for responses (15 seconds)...");
                await Task.Delay(15000);
                
                _cts.Cancel();
                await consumerTask;
            }
            finally
            {
                Shutdown();
                PrintSummary();
            }
            
            Console.WriteLine("\nTest complete.");
        }
        
        private void Connect()
        {
            Console.WriteLine("Connecting to IBM MQ...");
            
            var properties = new Hashtable
            {
                { MQConstants.HOST_NAME_PROPERTY, HostName },
                { MQConstants.PORT_PROPERTY, Port },
                { MQConstants.CHANNEL_PROPERTY, ChannelName },
                { MQConstants.TRANSPORT_PROPERTY, MQConstants.TRANSPORT_MQSERIES_MANAGED }
            };
            
            _queueManager = new MQQueueManager(QueueManagerName, properties);
            
            _requestQueue = _queueManager.AccessQueue(
                RequestQueueName,
                MQConstants.MQOO_OUTPUT | MQConstants.MQOO_FAIL_IF_QUIESCING
            );
            
            _responseQueue = _queueManager.AccessQueue(
                ResponseQueueName,
                MQConstants.MQOO_INPUT_AS_Q_DEF | MQConstants.MQOO_FAIL_IF_QUIESCING
            );
            
            Console.WriteLine("✓ Connected to IBM MQ");
        }
        
        private async Task StartResponseConsumerAsync(CancellationToken ct)
        {
            Console.WriteLine("✓ Response consumer started");
            
            await Task.Run(() =>
            {
                var gmo = new MQGetMessageOptions
                {
                    Options = MQConstants.MQGMO_WAIT | MQConstants.MQGMO_FAIL_IF_QUIESCING,
                    WaitInterval = 1000
                };
                
                while (!ct.IsCancellationRequested)
                {
                    try
                    {
                        var message = new MQMessage();
                        _responseQueue!.Get(message, gmo);
                        
                        var body = message.ReadString(message.MessageLength);
                        ProcessMessage(body);
                    }
                    catch (MQException ex) when (ex.ReasonCode == MQConstants.MQRC_NO_MSG_AVAILABLE)
                    {
                        // Normal timeout, continue
                    }
                    catch (Exception ex)
                    {
                        if (!ct.IsCancellationRequested)
                        {
                            Console.Error.WriteLine($"Error receiving message: {ex.Message}");
                        }
                    }
                }
            }, ct);
        }
        
        private void ProcessMessage(string body)
        {
            try
            {
                Console.WriteLine();
                Console.WriteLine("════════════════════════════════════════════════════════════");
                Console.WriteLine("📥 RESPONSE RECEIVED");
                
                var response = JsonNode.Parse(body);
                if (response == null) return;
                
                var requestNode = response["request"];
                var responseNode = response["response"];
                
                var messageId = requestNode?["message_id"]?.GetValue<string>() ?? "unknown";
                var success = responseNode?["success"]?.GetValue<bool>() ?? false;
                var operation = requestNode?["operation"]?.GetValue<string>() ?? "unknown";
                var processingTime = response["processing_time_ms"]?.GetValue<long>() ?? 0;
                
                Console.WriteLine($"   Message ID: {messageId}");
                Console.WriteLine($"   Success: {success}");
                Console.WriteLine($"   Operation: {operation}");
                Console.WriteLine($"   Processing Time: {processingTime}ms");
                
                if (success)
                {
                    var result = responseNode?["result"]?.ToJsonString();
                    if (!string.IsNullOrEmpty(result))
                    {
                        Console.WriteLine(result.Length > 100 
                            ? $"   Result: {result[..100]}..." 
                            : $"   Result: {result}");
                    }
                }
                else
                {
                    var errorCode = responseNode?["error_code"]?.GetValue<string>() ?? "UNKNOWN";
                    var error = responseNode?["error"]?.GetValue<string>() ?? "Unknown error";
                    Console.WriteLine($"   Error Code: {errorCode}");
                    Console.WriteLine($"   Error: {error}");
                }
                Console.WriteLine("════════════════════════════════════════════════════════════");
                
                lock (_responseLock)
                {
                    _receivedResponses.Add(response);
                }
                
                if (_pendingRequests.TryRemove(messageId, out var tcs))
                {
                    tcs.SetResult(response);
                }
            }
            catch (Exception ex)
            {
                Console.Error.WriteLine($"Error processing message: {ex.Message}");
            }
        }
        
        private async Task RunTestSuiteAsync()
        {
            Console.WriteLine("\n════════════════════════════════════════════════════════════");
            Console.WriteLine("STARTING TEST SUITE");
            Console.WriteLine("════════════════════════════════════════════════════════════");
            
            // Test 1: SET
            Console.WriteLine("\n--- Test 1: SET operation ---");
            var setRequest = CreateRequest("SET", TestRegion, "ibmmq-test:user:1001");
            setRequest["value"] = new JsonObject
            {
                ["name"] = "John Doe",
                ["email"] = "john@example.com",
                ["broker"] = "ibmmq"
            };
            setRequest["ttl"] = 300;
            PublishRequest(setRequest);
            await Task.Delay(1000);
            
            // Test 2: GET
            Console.WriteLine("\n--- Test 2: GET operation ---");
            PublishRequest(CreateRequest("GET", TestRegion, "ibmmq-test:user:1001"));
            await Task.Delay(1000);
            
            // Test 3: EXISTS
            Console.WriteLine("\n--- Test 3: EXISTS operation ---");
            PublishRequest(CreateRequest("EXISTS", TestRegion, "ibmmq-test:user:1001"));
            await Task.Delay(1000);
            
            // Test 4: SET another
            Console.WriteLine("\n--- Test 4: SET another key ---");
            var setRequest2 = CreateRequest("SET", TestRegion, "ibmmq-test:user:1002");
            setRequest2["value"] = new JsonObject
            {
                ["name"] = "Jane Smith",
                ["role"] = "admin"
            };
            PublishRequest(setRequest2);
            await Task.Delay(1000);
            
            // Test 5: KEYS
            Console.WriteLine("\n--- Test 5: KEYS operation ---");
            PublishRequest(CreateRequest("KEYS", TestRegion, "ibmmq-test:*"));
            await Task.Delay(1000);
            
            // Test 6: DELETE
            Console.WriteLine("\n--- Test 6: DELETE operation ---");
            PublishRequest(CreateRequest("DELETE", TestRegion, "ibmmq-test:user:1002"));
            await Task.Delay(1000);
            
            // Test 7: GET non-existent
            Console.WriteLine("\n--- Test 7: GET non-existent key ---");
            PublishRequest(CreateRequest("GET", TestRegion, "ibmmq-test:nonexistent:key"));
            await Task.Delay(1000);
            
            Console.WriteLine("\n✓ All requests sent");
        }
        
        private JsonObject CreateRequest(string operation, string region, string key)
        {
            var messageId = $"req-{Guid.NewGuid().ToString()[..12]}";
            
            var request = new JsonObject
            {
                ["operation"] = operation,
                ["region"] = region,
                ["key"] = key,
                ["api_key"] = ApiKey,
                ["message_id"] = messageId
            };
            
            _pendingRequests[messageId] = new TaskCompletionSource<JsonNode>();
            return request;
        }
        
        private void PublishRequest(JsonObject request)
        {
            var messageId = request["message_id"]!.GetValue<string>();
            var requestJson = request.ToJsonString();
            
            Console.WriteLine("📤 SENDING REQUEST");
            Console.WriteLine($"   Message ID: {messageId}");
            Console.WriteLine($"   Operation: {request["operation"]}");
            Console.WriteLine($"   Key: {request["key"]}");
            
            var message = new MQMessage();
            message.WriteString(requestJson);
            message.Format = MQConstants.MQFMT_STRING;
            message.Persistence = MQConstants.MQPER_PERSISTENT;
            
            var pmo = new MQPutMessageOptions();
            _requestQueue!.Put(message, pmo);
            
            Console.WriteLine("   ✓ Sent");
        }
        
        private void Shutdown()
        {
            Console.WriteLine("\nShutting down...");
            
            try
            {
                _requestQueue?.Close();
                _responseQueue?.Close();
                _queueManager?.Disconnect();
            }
            catch (Exception ex)
            {
                Console.Error.WriteLine($"Error closing connections: {ex.Message}");
            }
        }
        
        private void PrintSummary()
        {
            Console.WriteLine("\n════════════════════════════════════════════════════════════");
            Console.WriteLine("TEST SUMMARY");
            Console.WriteLine("════════════════════════════════════════════════════════════");
            
            lock (_responseLock)
            {
                Console.WriteLine($"Total responses received: {_receivedResponses.Count}");
            }
            Console.WriteLine($"Pending requests (no response): {_pendingRequests.Count}");
            
            if (!_pendingRequests.IsEmpty)
            {
                Console.WriteLine("\nRequests without responses:");
                foreach (var messageId in _pendingRequests.Keys)
                {
                    Console.WriteLine($"  - {messageId}");
                }
            }
            
            int successCount, errorCount;
            lock (_responseLock)
            {
                successCount = _receivedResponses.Count(r =>
                {
                    var responseNode = r["response"];
                    return responseNode?["success"]?.GetValue<bool>() ?? false;
                });
                errorCount = _receivedResponses.Count - successCount;
            }
            
            Console.WriteLine($"\nSuccessful operations: {successCount}");
            Console.WriteLine($"Failed operations: {errorCount}");
            Console.WriteLine("════════════════════════════════════════════════════════════");
        }
    }
}
